🎉  New destination: databend

hantmac opened the pull request:
## What

- Add new databend destination

## How
- Implemented this destination connector using [Databend official driver](https://github.com/databendcloud/databend-py) and [databend sqlalchemy](https://github.com/databendcloud/databend-sqlalchemy).
- Upstream [dbt databend adapter](https://github.com/databendcloud/dbt-databend/) has run pass the [example](https://github.com/dbt-labs/jaffle_shop) and more detail in this [wiki](https://github.com/databendcloud/dbt-databend/wiki/How-to-use-dbt-with-Databend-Cloud)
## Recommended reading order
1. `docs/integrations/destinations/databend.md`
3. `spec.json`
4. `destination_databend/destination.py`
5. `destination_databend/client.py`
6.  `destination_databend/writer.py`

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<summary><strong>Unit</strong></summary>

<img width="1105" alt="image" src="https://user-images.githubusercontent.com/7600925/204121263-3fdc245e-8a8e-4813-a299-4ee5acf42afd.png">

<img width="1516" alt="image" src="https://user-images.githubusercontent.com/7600925/205044796-57a1f9db-f907-42c7-bb26-935cc6540ec9.png">


<summary><strong>Integration</strong></summary>

<img width="1135" alt="image" src="https://user-images.githubusercontent.com/7600925/204231433-6f2b6a8f-34b3-4e1a-92f8-5708e339c3d6.png">

hantmac committed:
feat: Add databend destination

hantmac committed:
docs

hantmac committed:
fix

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=19815) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=19815) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=19815) it.</sub>

hantmac committed:
Merge branch 'master' into feat/databend-destination

hantmac committed:
fix

hantmac committed:
fix tests

hantmac committed:
support normalization for databend

hantmac committed:
update

hantmac committed:
fix gradle build

hantmac committed:
fix

hantmac committed:
update

hantmac committed:
fix transform

hantmac committed:
Merge branch 'master' into feat/databend-destination

hantmac committed:
fix spec.json

natalyjazzviolin committed:
Merge branch 'master' into feat/databend-destination

hantmac commented:
@itaseskii  Hi bro, how has the progress on this PR been recently? 😄

itaseskii commented:
/test connector=connectors/destination-databend

itaseskii commented on airbyte-integrations/connectors/destination-databend/destination_databend/writer.py:
```diff
@@ -0,0 +1,127 @@
+
+class DatabendSQLWriter(DatabendWriter):
+    """
+    Data writer using the SQL writing strategy. Data is buffered in memory
```
why is buffering referred to as `less effective strategy` in this comment?

btw when buffering records a good practice is to take into consideration the actual size of the data in the buffer as well.

hantmac replied:
This comment is not correct. I remove it.

itaseskii commented on airbyte-integrations/connectors/destination-databend/integration_tests/integration_test.py:
```diff
@@ -0,0 +1,159 @@
+#
```
Have you considered relying on the destination acceptance tests as a baseline for correctness of your connector? They test a lot more scenarios than covered here. They will require some Java coding on your side though.

Do the current tests assume there is a running Databend instance to connect to? If so I suggest running a databend docker container as part of the tests https://github.com/testcontainers/testcontainers-python

hantmac replied:
Ok, I will consider the destination acceptance tests as a baseline and would you like to give me some docs about writing the tests code? Or I can add the test cases later because I have already successfully run the airbyte + dbt normalization pipeline.
![image](https://user-images.githubusercontent.com/7600925/208065401-b1e52d1a-e39c-4c3a-bb5d-e049fa34adb6.png)

Another question, the current test requires a connection string to connect to databend cloud. So I don't konw how to provide it in tests and I wonder how snowflake mock there client and I will do a survey on it🤔.

itaseskii commented on airbyte-integrations/connectors/destination-databend/destination_databend/writer.py:
```diff
@@ -0,0 +1,127 @@
+
+        :param name: table name to delete.
+        """
+        cursor = self.cursor
```
why is this assignment needed cant you invoke `self.cursor.execute(f"DROP TABLE IF EXISTS _airbyte_raw_{name}")` directly?

hantmac replied:
Thanks a lot. 😄

hantmac commented:
> > @itaseskii Hi bro, how has the progress on this PR been recently? smile
> 
> Hi @hantmac check my comments

Thanks for your work! I will check the comments. Thanks again!

hantmac committed:
fix

hantmac committed:
more check

hantmac committed:
add finally for connector

hantmac committed:
fix conflicts

hantmac commented on airbyte-integrations/connectors/destination-databend/destination_databend/writer.py:
```diff
@@ -0,0 +1,127 @@
+            to the databse.
+        """
+        self.client = client
+        self.cursor = client.open()
```
yse, it should it be closed in a finally block

hantmac commented on airbyte-integrations/connectors/destination-databend/destination_databend/writer.py:
```diff
@@ -0,0 +1,127 @@
+        """
+        super().__init__(client)
+
+    def _flush(self) -> None:
```
The logic for the flush implementation is in python driver for databend, https://github.com/databendcloud/databend-py/blob/main/databend_py/client.py#L106.

Yes, all the queries in the DatabendWriter are SQL queries as well for now and if we want to add another `writer`such as `S3Writer` we can extend the class `DatabendWriter `.

hantmac commented on airbyte-integrations/connectors/destination-databend/destination_databend/writer.py:
```diff
@@ -0,0 +1,127 @@
+        # id, written_at, data
+        for table, data in self._buffer.items():
+            cursor.execute(
+                f"INSERT INTO _airbyte_raw_{table} (_airbyte_ab_id,_airbyte_emitted_at,_airbyte_data) VALUES (%, %, %)",
```
Of course, here is the test code https://github.com/databendcloud/databend-py/blob/main/tests/test_client.py#L49 and the logic of `batch insert` in databend driver https://github.com/databendcloud/databend-py/blob/main/databend_py/client.py#L106.

hantmac committed:
fix comment

hantmac committed:
fix

hantmac commented:
> > @itaseskii Hi bro, how has the progress on this PR been recently? smile
> 
> Hi @hantmac check my comments

Hi @itaseskii I have resolved most of your comments please have a look. 😄

itaseskii commented:
> > > @itaseskii Hi bro, how has the progress on this PR been recently? smile
> > 
> > 
> > Hi @hantmac check my comments
> 
> Hi @itaseskii I have resolved your comments please have a look. smile

@hantmac please don't resolve comments until verified.

hantmac commented:
> > > > @itaseskii Hi bro, how has the progress on this PR been recently? smile
> > > 
> > > 
> > > Hi @hantmac check my comments
> > 
> > 
> > Hi @itaseskii I have resolved your comments please have a look. smile
> 
> @hantmac please don't resolve comments until verified.

Sorry，it's my fault.

hantmac committed:
Merge branch 'master' into feat/databend-destination

hantmac committed:
Merge branch 'master' into feat/databend-destination

hantmac commented on airbyte-integrations/connectors/destination-databend/destination_databend/destination.py:
```diff
@@ -0,0 +1,86 @@
+        try:
+            client = DatabendClient(**config)
+            cursor = client.open()
+            cursor.execute("select 1")
```
Make sense and fixed.

grishick committed:
Merge branch 'master' into feat/databend-destination

grishick reviewed (changes requested):
@hantmac we recently removed `NormalizationRunnerFactory` and moved specification of normalization into `destination_definitions.yaml`. In order to add normalization support, you would need to add a section to `destination_definitions.yaml` for the new connector. It could look like following (note: generate a random UUID v4 for `destinationDefinitionId`):
```
name: Databend
  destinationDefinitionId: aa0cf6d4-fa7a-4d4a-9a42-6ddb5687d28a
  dockerRepository: airbyte/destination-databend
  dockerImageTag: 0.1.0
  documentationUrl: https://docs.airbyte.com/integrations/destinations/databend
  normalizationConfig:
    normalizationRepository: airbyte/normalization-databend
    normalizationTag: 0.2.25
    normalizationIntegrationType: databend
  supportsDbt: true
  releaseStage: alpha
```

hantmac committed:
add section to destination_definitions

hantmac commented:
> @hantmac we recently removed `NormalizationRunnerFactory` and moved specification of normalization into `destination_definitions.yaml`. In order to add normalization support, you would need to add a section to `destination_definitions.yaml` for the new connector. It could look like following (note: generate a random UUID v4 for `destinationDefinitionId`):
> 
> ```
> name: Databend
>   destinationDefinitionId: aa0cf6d4-fa7a-4d4a-9a42-6ddb5687d28a
>   dockerRepository: airbyte/destination-databend
>   dockerImageTag: 0.1.0
>   documentationUrl: https://docs.airbyte.com/integrations/destinations/databend
>   normalizationConfig:
>     normalizationRepository: airbyte/normalization-databend
>     normalizationTag: 0.2.25
>     normalizationIntegrationType: databend
>   supportsDbt: true
>   releaseStage: alpha
> ```

Hi @grishick, thanks for your review. I have added a section to `destination_definitions.yaml`.

hantmac committed:
default port

hantmac committed:
add databend icon

marcosmarxm commented:
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

grishick commented on airbyte-integrations/connectors/destination-databend/README.md:
```diff
@@ -0,0 +1,123 @@
+**If you are a community contributor**, follow the instructions in the [documentation](https://docs.airbyte.io/integrations/destinations/databend)
+to generate the necessary credentials. Then create a file `secrets/config.json` conforming to the `destination_databend/spec.json` file.
+Note that the `secrets` directory is gitignored by default, so there is no danger of accidentally checking in sensitive information.
+See `integration_tests/sample_config.json` for a sample config file.
```
Could you please add `sample_config.json` ?

hantmac replied:
Done. Thank u very much!

hantmac committed:
add sample_config.json

hantmac committed:
Merge branch 'master' into feat/databend-destination

grishick commented:
I will try to move this PR the last mile here: https://github.com/airbytehq/airbyte/pull/20909

grishick commented:
I got a Databend cloud account set up, I added the credentials to `secrets/config.json` and tried running integration tests as
```
./gradlew :airbyte-integrations:connectors:destination-databend:integrationTest
```

I got the following error:
```
         ============================= test session starts ==============================
         platform darwin -- Python 3.9.11, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /Users/gregsolovyev/git/airbyte/airbyte-integrations/connectors/destination-databend/.venv/bin/python
         cachedir: .pytest_cache
         rootdir: /Users/gregsolovyev/git/airbyte/airbyte-integrations/connectors/destination-databend/integration_tests, configfile: ../../../../pyproject.toml
         collecting ... collected 3 items
         
         integration_tests/integration_test.py::test_check_valid_config {"type": "DEBUG", "message": "http sql: DROP TABLE IF EXISTS test", "data": {}}
         {"type": "DEBUG", "message": "http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"POST /v1/query/ HTTP/1.1\" 200 616", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"GET /v1/query/8bd905bd-8eb8-4732-9ff2-f5d7aeb40225/final HTTP/1.1\" 200 568", "data": {}}
         {"type": "DEBUG", "message": "http sql: CREATE TABLE if not exists test (x Int32,y VARCHAR)", "data": {}}
         {"type": "DEBUG", "message": "http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"POST /v1/query/ HTTP/1.1\" 200 616", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"GET /v1/query/7b7cec26-7d75-42a3-82b0-87457e98610a/final HTTP/1.1\" 200 568", "data": {}}
         FAILED
         integration_tests/integration_test.py::test_check_invalid_config PASSED
         integration_tests/integration_test.py::test_write {"type": "LOG", "log": {"level": "INFO", "message": "Using the SQL writing strategy"}}
         {"type": "DEBUG", "message": "http sql: \n        CREATE TABLE IF NOT EXISTS _airbyte_raw_append_stream (\n            _airbyte_ab_id TEXT,\n            _airbyte_emitted_at TIMESTAMP,\n            _airbyte_data TEXT\n        )\n        ", "data": {}}
         {"type": "DEBUG", "message": "http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"POST /v1/query/ HTTP/1.1\" 200 615", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"GET /v1/query/c88c416b-1cd8-4874-88d3-fe49871f28e9/final HTTP/1.1\" 200 567", "data": {}}
         {"type": "DEBUG", "message": "http sql: DROP TABLE IF EXISTS _airbyte_raw_overwrite_stream", "data": {}}
         {"type": "DEBUG", "message": "http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"POST /v1/query/ HTTP/1.1\" 200 616", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"GET /v1/query/2f997911-fd0a-4a7d-bd22-9880153c2b63/final HTTP/1.1\" 200 568", "data": {}}
         {"type": "LOG", "log": {"level": "INFO", "message": "Stream overwrite_stream is wiped."}}
         {"type": "DEBUG", "message": "http sql: \n        CREATE TABLE IF NOT EXISTS _airbyte_raw_overwrite_stream (\n            _airbyte_ab_id TEXT,\n            _airbyte_emitted_at TIMESTAMP,\n            _airbyte_data TEXT\n        )\n        ", "data": {}}
         {"type": "DEBUG", "message": "http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"POST /v1/query/ HTTP/1.1\" 200 624", "data": {}}
         {"type": "DEBUG", "message": "Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443", "data": {}}
         {"type": "DEBUG", "message": "https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 \"GET /v1/query/368e3d1b-cfda-4a73-a3d1-cf0532aeedc9/final HTTP/1.1\" 200 576", "data": {}}
         FAILED
         
         =================================== FAILURES ===================================
         ___________________________ test_check_valid_config ____________________________
         
         config = {'database': 'default', 'host': 'tnibtyx08--default.ch.aws-us-east-2.default.databend.com', 'password': '******', 'port': 443, ...}
         
             def test_check_valid_config(config: Mapping):
                 outcome = DestinationDatabend().check(logging.getLogger('airbyte'), config)
         >       assert outcome.status == Status.SUCCEEDED
         E       AssertionError: assert <Status.FAILED: 'FAILED'> == <Status.SUCCEEDED: 'SUCCEEDED'>
         E         +<Status.FAILED: 'FAILED'>
         E         -<Status.SUCCEEDED: 'SUCCEEDED'>
         
         integration_tests/integration_test.py:66: AssertionError
         ------------------------------ Captured log call -------------------------------
         DEBUG    databend_py.log:connection.py:109 http sql: DROP TABLE IF EXISTS test
         DEBUG    databend_py.log:connection.py:113 http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "POST /v1/query/ HTTP/1.1" 200 616
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "GET /v1/query/8bd905bd-8eb8-4732-9ff2-f5d7aeb40225/final HTTP/1.1" 200 568
         DEBUG    databend_py.log:connection.py:109 http sql: CREATE TABLE if not exists test (x Int32,y VARCHAR)
         DEBUG    databend_py.log:connection.py:113 http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "POST /v1/query/ HTTP/1.1" 200 616
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "GET /v1/query/7b7cec26-7d75-42a3-82b0-87457e98610a/final HTTP/1.1" 200 568
         __________________________________ test_write __________________________________
         
         config = {'database': 'default', 'host': 'tnibtyx08--default.ch.aws-us-east-2.default.databend.com', 'password': 'cjgvdx5ee5xe', 'port': 443, ...}
         configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='append_stream', json_schema={'typ...cremental'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])
         client = <destination_databend.client.DatabendClient object at 0x104958d30>
         
             def test_write(config: Mapping, configured_catalog: ConfiguredAirbyteCatalog, client: DatabendClient):
                 """
                 This test verifies that:
                     1. writing a stream in "overwrite" mode overwrites any existing data for that stream
                     2. writing a stream in "append" mode appends new records without deleting the old ones
                     3. The correct state message is output by the connector at the end of the sync
                 """
                 append_stream, overwrite_stream = configured_catalog.streams[0].stream.name, configured_catalog.streams[1].stream.name
                 first_state_message = _state({"state": "1"})
                 first_record_chunk = [_record(append_stream, str(i), i) for i in range(5)] + [_record(overwrite_stream, str(i), i) for i in range(5)]
             
                 second_state_message = _state({"state": "2"})
                 second_record_chunk = [_record(append_stream, str(i), i) for i in range(5, 10)] + [
                     _record(overwrite_stream, str(i), i) for i in range(5, 10)
                 ]
             
                 destination = DestinationDatabend()
             
                 expected_states = [first_state_message, second_state_message]
         >       output_states = list(
                     destination.write(
                         config, configured_catalog, [*first_record_chunk, first_state_message, *second_record_chunk, second_state_message]
                     )
                 )
         
         integration_tests/integration_test.py:125: 
         _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
         destination_databend/destination.py:66: in write
             writer.flush()
         destination_databend/writer.py:128: in flush
             self._flush()
         destination_databend/writer.py:117: in _flush
             cursor.execute(
         .venv/lib/python3.9/site-packages/databend_sqlalchemy/connector.py:176: in execute
             sql = operation % _escaper.escape_args(parameters)
         .venv/lib/python3.9/site-packages/databend_sqlalchemy/connector.py:33: in escape_args
             return tuple(self.escape_item(x) for x in parameters)
         .venv/lib/python3.9/site-packages/databend_sqlalchemy/connector.py:33: in <genexpr>
             return tuple(self.escape_item(x) for x in parameters)
         _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
         
         self = <databend_sqlalchemy.connector.ParamEscaper object at 0x10459a9d0>
         item = 'db7a9b64-fa0e-4b50-b032-c682f30a66fd'
         
             def escape_item(self, item):
                 if item is None:
                     return 'NULL'
                 elif isinstance(item, (int, float)):
                     return self.escape_number(item)
                 elif isinstance(item, datetime):
                     return self.escape_string(item.strftime("%Y-%m-%d %H:%M:%S"))
                 else:
         >           raise Exception("Unsupported object {}".format(item))
         E           Exception: Unsupported object db7a9b64-fa0e-4b50-b032-c682f30a66fd
         
         .venv/lib/python3.9/site-packages/databend_sqlalchemy/connector.py:57: Exception
         ------------------------------ Captured log call -------------------------------
         INFO     airbyte:writer.py:132 Using the SQL writing strategy
         DEBUG    databend_py.log:connection.py:109 http sql: 
                 CREATE TABLE IF NOT EXISTS _airbyte_raw_append_stream (
                     _airbyte_ab_id TEXT,
                     _airbyte_emitted_at TIMESTAMP,
                     _airbyte_data TEXT
                 )
                 
         DEBUG    databend_py.log:connection.py:113 http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "POST /v1/query/ HTTP/1.1" 200 615
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "GET /v1/query/c88c416b-1cd8-4874-88d3-fe49871f28e9/final HTTP/1.1" 200 567
         DEBUG    databend_py.log:connection.py:109 http sql: DROP TABLE IF EXISTS _airbyte_raw_overwrite_stream
         DEBUG    databend_py.log:connection.py:113 http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "POST /v1/query/ HTTP/1.1" 200 616
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "GET /v1/query/2f997911-fd0a-4a7d-bd22-9880153c2b63/final HTTP/1.1" 200 568
         INFO     airbyte:destination.py:50 Stream overwrite_stream is wiped.
         DEBUG    databend_py.log:connection.py:109 http sql: 
                 CREATE TABLE IF NOT EXISTS _airbyte_raw_overwrite_stream (
                     _airbyte_ab_id TEXT,
                     _airbyte_emitted_at TIMESTAMP,
                     _airbyte_data TEXT
                 )
                 
         DEBUG    databend_py.log:connection.py:113 http headers {'Content-Type': 'application/json', 'Accept': 'application/json', 'X-DATABEND-ROUTE': 'warehouse', 'Authorization': 'Basic Y2xvdWRhcHA6Y2pndmR4NWVlNXhl'}
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "POST /v1/query/ HTTP/1.1" 200 624
         DEBUG    urllib3.connectionpool:connectionpool.py:1003 Starting new HTTPS connection (1): tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443
         DEBUG    urllib3.connectionpool:connectionpool.py:456 https://tnibtyx08--default.ch.aws-us-east-2.default.databend.com:443 "GET /v1/query/368e3d1b-cfda-4a73-a3d1-cf0532aeedc9/final HTTP/1.1" 200 576
         =========================== short test summary info ============================
         FAILED integration_tests/integration_test.py::test_check_valid_config - Asser...
         FAILED integration_tests/integration_test.py::test_write - Exception: Unsuppo...
         ========================= 2 failed, 1 passed in 5.46s ==========================

> Task :airbyte-integrations:connectors:destination-databend:_customIntegrationTestsCoverage FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':airbyte-integrations:connectors:destination-databend:_customIntegrationTestsCoverage'.
> Python call failed: .venv/bin/python -m coverage run --data-file=integration_tests/.coverage.customIntegrationTests --rcfile=/Users/gregsolovyev/git/airbyte/pyproject.toml -m pytest -s integration_tests -c /Users/gregsolovyev/git/airbyte/pyproject.toml

```

grishick commented:
It appears that escape_item is unable to process a string parameter that looks like UUID:

```
    def escape_item(self, item):
        if item is None:
            return 'NULL'
        elif isinstance(item, (int, float)):
            return self.escape_number(item)
        elif isinstance(item, datetime):
            return self.escape_string(item.strftime("%Y-%m-%d %H:%M:%S"))
        else:
            raise Exception("Unsupported object {}".format(item))
```

hantmac commented:
> It appears that `ParamEscaper::escape_item` is unable to process a string parameter that looks like UUID:
> 
> ```
>     def escape_item(self, item):
>         if item is None:
>             return 'NULL'
>         elif isinstance(item, (int, float)):
>             return self.escape_number(item)
>         elif isinstance(item, datetime):
>             return self.escape_string(item.strftime("%Y-%m-%d %H:%M:%S"))
>         else:
>             raise Exception("Unsupported object {}".format(item))
> ```

Thanks for your help. The `databend_sqlalchemy` has bug and have been fixed. After that the integration tests passed.

```
collected 3 items                                                                                                                                                                                             

integration_tests/integration_test.py::test_check_valid_config 
PASSED
integration_tests/integration_test.py::test_check_invalid_config PASSED
integration_tests/integration_test.py::test_write {"type": "LOG", "log": {"level": "INFO", "message": "Using the SQL writing strategy"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Stream overwrite_stream is wiped."}}
{"type": "LOG", "log": {"level": "INFO", "message": "Using the SQL writing strategy"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Stream overwrite_stream is wiped."}}
PASSED
```

hantmac committed:
Merge branch 'master' into feat/databend-destination

grishick commented:
I added secrets to GSM and am running integration tests here: https://github.com/airbytehq/airbyte/pull/20909

grishick commented:
@hantmac I added secrets to secrets manager and ran integration tests via CI. Destination tests pass (🎉 ), but normalization tests fail: https://github.com/airbytehq/airbyte/actions/runs/3839797452/jobs/6538073823

grishick commented:
This contribution has been merged in this PR: https://github.com/airbytehq/airbyte/pull/20909 w/o normalization support
@hantmac feel free to followup with adding normalization support
